packages/block-editor/src/hooks/block-bindings.js

```diff
@@ -192,50 +190,51 @@ export const BlockBindingsPanel = ( { name: blockName, metadata } ) => {
 	const bindableAttributes = getBindableAttributes( blockName );
 	const dropdownMenuProps = useToolsPanelDropdownMenuProps();
 
-	// `useSelect` is used purposely here to ensure `getFieldsList`
-	// is updated whenever there are updates in block context.
-	// `source.getFieldsList` may also call a selector via `registry.select`.
-	const _fieldsList = {};
-	const { fieldsList, canUpdateBlockBindings } = useSelect(
-		( select ) => {
-			if ( ! bindableAttributes || bindableAttributes.length === 0 ) {
-				return EMPTY_OBJECT;
-			}
-			const { getBlockBindingsSources } = unlock( blocksPrivateApis );
-			const registeredSources = getBlockBindingsSources();
-			Object.entries( registeredSources ).forEach(
-				( [ sourceName, { getFieldsList, usesContext } ] ) => {
-					if ( getFieldsList ) {
-						// Populate context.
-						const context = {};
-						if ( usesContext?.length ) {
-							for ( const key of usesContext ) {
-								context[ key ] = blockContext[ key ];
-							}
-						}
-						const sourceList = getFieldsList( {
-							registry,
-							context,
-						} );
-						// Only add source if the list is not empty.
-						if ( Object.keys( sourceList || {} ).length ) {
-							_fieldsList[ sourceName ] = { ...sourceList };
+	const { canUpdateBlockBindings } = useSelect( ( select ) => {
+		return {
+			canUpdateBlockBindings:
+				select( blockEditorStore ).getSettings().canUpdateBlockBindings,
+		};
+	}, [] );
+
+	/**
+	 * Create new selector for fieldsList to avoid unnecessary re-renders.
+	 * See: https://github.com/WordPress/gutenberg/pull/64072#discussion_r1764693730
+	 *
+	 * `useSelect` is used purposely here to ensure `getFieldsList` is updated
+	 * whenever there are updates in block context.
+	 * `source.getFieldsList` may also call a selector via `registry.select`.
+	 */
+	const fieldsList = useSelect( () => {
```

Comment:
As suggested here, I moved by @jsnajdr [here](https://github.com/WordPress/gutenberg/pull/64072#discussion_r1771409849), I return the `fieldsList` as the top-level object in the useSelect result.

However, this triggers the `useSelect` warning. After triaging it a bit, I believe it is caused because `fieldsList` is an object of objects, and when we check `isShallowEqual` [here](https://github.com/WordPress/gutenberg/blob/9fef8fb128ef280ed5348c4f4ba3ec93c883fe6a/packages/data/src/components/use-select/index.js#L161), it returns false. I guess that is expected? Should we keep the variable outside or should this be fixed in another way?

Reply:
Oh, that's because the values in `fieldsList`, the results of `getFieldsList` calls, are also not memoized 🙁 Every time `getFieldsList` is called, it returns a new value. I wonder if anything can be done about that?

Memoization and immutability is sometimes very hard to achieve, it's a weakness of Redux-like state management.

Reply:
I'm still trying to understand better how memoization works 😄 Let me ask some questions to see how I can proceed.

I was testing different scenarios to understand it better. Removing the `getFieldsList` calls to avoid interferences, I added a `useSelect` returning some objects.

```js
const fieldsList = useSelect( () => {
    // This triggers the warning.
    return {
	test: {
            subprop: 'subprop',
	},
    };

    // This DOES NOT trigger the warning.
    return { test: 'prop' };
}, [] );
```

Is this expected? Does this mean we have to memoized the value we are returning independently of what `getFieldsList` does?

Reply:
I was triaging this a bit more and in my testing if the `useSelect` returns an object of objects, the warning is always there.

I'm not saying this is the right solution, but for testing, I checked that running [here](https://github.com/WordPress/gutenberg/blob/93323fd1a3cecd0771ead22b16a9661c53be6212/packages/data/src/components/use-select/index.js#L161) `fastDeepEqual` instead of `isShallowEqual` doesn't trigger the warning. But I assume we are using `isShallowEqual` for a good reason?

Any idea how this could be solved? I tried creating a ref with `useRef` or memorizing the `fieldsList`, but it doesn't seem to work.

Reply:
The principle of function memoization is that when a function has the same inputs, it should return the same output. In a sense that the old value is `===` to the other. This function is not memoized:
```js
function getFieldsList() {
  return { test: 'prop' };
}
```
Although it always returns the same thing, it actually doesn't return the same thing: it's a different object on each call. If such a return value is passed as an `useEffect` dependency, or as a prop, or as context value, it tells React that something has changed (not `===` to the previous) value while in fact nothing has changed. That's bad, because it triggers a lot of extra work and can also lead to bugs.

That's why we need to invest a lot of care into memoizing return values and why sometimes library code issues warnings when it detects problems. JavaScript unfortunately doesn't do it very well natively.

The return value of `useSelect` is a "container of results". It's not treated as a "result" itself. That's why it can be a different object on each call. Because it's not supposed to be used as "result" itself, i.e., passed as a prop etc. Only it's members are "results".

Returning `{ test: 'prop' }` is OK because the real result is the `'prop'` string and it's always `===` to any other `'prop'` string. It's different with objects and arrays. In `{ test: { subprop: 'subprop' } }` the result is the `{ subprop: 'subprop' }` object. It's a different object on each call and that's why `useSelect` complains.

Ideally the `getFieldsList` functions should return constant objects, or, if the result is dynamic and depends on something, use some cache. There are many utils that help with that, like `memize`.

Reply:
Thanks a lot for the detailed explanation, it is really useful ☺️ 

I don't think we can return constant objects from `getFieldsList` because they might depend on the store. For example, in case of post meta, we are calling `getEditedEntityRecord`. And ideally, it should reflect the changes made in the client. Other sources could have different use cases in the future.

I've tried to cache `getFieldsList` but I encountered some issues where the component isn't rerendered when the store changes, so I need to investigate it further.

Reply:
> I don't think we can return constant objects from `getFieldsList` because they might depend on the store.

In that case you should create a new constant object when the relevant store information changes. Memoization APIs usually have array of dependencies as arguments:
```js
createSelector(
  ( state ) => { // calculate the result
    return [ getA( state ), getB( state ) ];
  },
  ( state ) => [ getA( state ), getB( state ) ], // state the array of dependencies
)
```
or
```js
useMemo( () => {
  return [ a, b ];
}, [ a, b ] );
```
the point is that the first argument, the main function, is called only when dependencies change. Otherwise the cached value is returned.

In your case `getEditedEntityRecord` would be in dependencies in one form or another.

> I've tried to cache `getFieldsList` but I encountered some issues where the component isn't rerendered

What code are you testing with? In the Gutenberg repo itself I don't see any code that would actually create bindings, there is only the framework to provide this capability.

Reply:
To do things really 100% correctly, the `getPostMetaFields` function should be memoized. It depends on the meta fields registered for the post type and also on the `meta` field of a specific post. It can change only when one of these dependencies change. The code should look like:
```js
const getPostMetaFields = createSelector(
  ( registry, context ) => {
    // calculate the fields
  },
  ( registry, context ) => [ // calculate the dependencies
    registry.select.getEditedEntityRecord( context.postType, context.postId ).meta,
    registry.select.getRegisteredPostMeta( context.postType ),
  ]
);
```
I'm not sure if the `createSelector` implementation in `@wordpress/data` can really be used here. There are subtle details: the cache key for memoization includes not only the dependencies, but also the `context` argument, but if we call the `getFieldsList` function like:
```js
getFieldsList( registry, { postType, postId } )
```
then the `context` is always a different object, it's only its fields that are stable and should really be keys.

This is a very complex problem, memoization is very hard to do right.

Reply:
Thanks again for the detailed explanation 🙂 

I've been testing `createSelector` in this commit: https://github.com/WordPress/gutenberg/pull/65604/commits/e9e00d4f9b578680981e336857af925d654eaaeb. It seems to work fine. It reacts to the changes in the store, it doesn't trigger the warning of too many rerenders, and it solves some issues we were having before.

My main concern is that these APIs are supposed to be used by extenders (hopefully `getFieldsList` won't be part of WordPress 6.7), and asking them to use `createSelector` feels a bit weird to me. Is this something we could abstract somehow? Any feedback in this regard?

Apart from that, as you shared, I had to create a `context` constant outside of the `useSelect` because it was interpreting it was a different object each time. Not sure if there is a better way to handle that.